### PR TITLE
[mlir] Pass Options ownership modifications

### PR DESCRIPTION
**Pull request by nikalra**
This change makes two (related) changes: 

First, it updates the tablegen option for `ListOption` to emit a `SmallVector` instead of an `ArrayRef`. This brings `ListOption` more inline with the traditional `Option`, where values are typically provided using types that have storage. After this change, all options should be fully owned by a Pass' `Options` object after it has been fully constructed, unless the underlying type of the `Option` explicitly indicates otherwise.

Second, it updates the generated constructors for Passes to consume options by value instead of reference, and prefers moving options into the pass itself. This should be more efficient for non-trivial options objects, where the previous interface forced a copy to be materialized. Now, at worst case the API materializes a copy (no worse than before); at best-case, all options objects are moved into place. Ideally, we could update the Pass constructor to take an r-value reference to the Options object instead, but this approach will require numerous changes to existing passes and their factory functions.

### CODE REVIEW

File: mlir/unittests/TableGen/PassGenTest.cpp [hunk 1]
```diff
@@ -72,7 +72,7 @@ TEST(PassGenTest, PassOptions) {
   TestPassWithOptionsOptions options;
   options.testOption = 57;
 
-  llvm::SmallVector<int64_t, 2> testListOption = {1, 2};
+  llvm::SmallVector<int64_t> testListOption = {1, 2};
   options.testListOption = testListOption;
```
Comment (River707) [inline mlir/unittests/TableGen/PassGenTest.cpp:76]:
```suggestion
  options.testListOption = {1, 2};
```
Do you even need the temporary anymore?

File: mlir/include/mlir/Transforms/Passes.h [hunk 2]
```diff
@@ -15,6 +15,7 @@
 #define MLIR_TRANSFORMS_PASSES_H
 
 #include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
```
Comment (River707) [inline mlir/include/mlir/Transforms/Passes.h:18]: Why is PassManager.h being included now?

Comment (nikalra) [reply mlir/include/mlir/Transforms/Passes.h:18]: The inliner pass contains a vector of `OpPassManager` in `InlinerOptions`. After this change, the size of `OpPassManager` needs to be known to compile a TU containing this header.

**Review by River707 — APPROVED** [review]

File: mlir/lib/Dialect/Arith/Transforms/IntNarrowing.cpp [hunk 3]
```diff
@@ -757,7 +758,8 @@ struct ArithIntNarrowingPass final
     MLIRContext *ctx = op->getContext();
     RewritePatternSet patterns(ctx);
     populateArithIntNarrowingPatterns(
-        patterns, ArithIntNarrowingOptions{bitwidthsSupported});
+        patterns, ArithIntNarrowingOptions{SmallVector<unsigned>{
+                      bitwidthsSupported.begin(), bitwidthsSupported.end()}});
```
Comment (kuhar) [inline mlir/lib/Dialect/Arith/Transforms/IntNarrowing.cpp:762]: nit: `llvm::to_vector_of<unsigned>(bitwidthsSupported)`?

**Pull request merged**